Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/TrainControlSystem.cs

```diff
@@ -674,12 +674,15 @@ SignalFeatures NextGenericSignalFeatures(string signalFunctionTypeName, int item
                         if (head.Function == function)
                             functionHead = head;
                     signalTypeName = functionHead.SignalTypeName;
-                    foreach (var key in functionHead.signalType.DrawStates.Keys)
+                    if (functionHead?.signalType?.DrawStates != null)
                     {
-                        if (functionHead.signalType.DrawStates[key].Index == functionHead.draw_state)
+                        foreach (var key in functionHead.signalType.DrawStates.Keys)
                         {
-                            drawStateName = functionHead.signalType.DrawStates[key].Name;
-                            break;
+                            if (functionHead.signalType.DrawStates[key].Index == functionHead.draw_state)
+                            {
+                                drawStateName = functionHead.signalType.DrawStates[key].Name;
+                                break;
```

Comment:
Looks like we are out of sync with the guy reporting the error, he still run the yesterday morning version. This is his real line of error, 677. And here it is possible to be null, because if in the previous foreach no head was found with the function, the signalType will be default, which is null.

Reply:
I've reverted my first commit, and added this instead.

Reply:
But functionHead can't be null, because if it were null the SignalObject wouldn't be in playerTrainSignalList.

Reply:
No, that is not null, but var functionHead = default(SignalHead), and if (head.Function == function) not found, then functionHead.signalType stays default, which is null.

Reply:
Yes, but if (head.Function == function) should be always found, because otherwise the signal wouldn't be present in playerTrainSignalList

Reply:
But if it's not a valid signal type, then 
if (!Locomotive.Train.signalRef.SignalFunctions.ContainsKey(signalFunctionTypeName))
applies, which handles that case.

If the signal type is valid, then we get all signals which have a head of that type, which is
var playerTrainSignalList = Locomotive.Train.PlayerTrainSignals[dir][function];
Therefore, all signals in playerTrainSignalList always have at leas one matching signal type

Reply:
Then another idea: Still the signalType may be null. The SignalHead has 2 constructors, and only one of them set the signalType. The other one requires that to be set via the SetSignalType() method. But not all paths there set this. There is also a warning:

Trace.TraceWarning("SignalObject trItem={0}, trackNode={1} has SignalHead with undefined SignalType {2}.",
              mainSignal.trItem, mainSignal.trackNode, sigItem.SignalType);

This would needed to be searched for in the log file, which unfortunately we don't have. But if it is there, that means the signalType is null.

Reply:
Fair enough, signalType can be null if there is a mismatch between TDB/world and SIGCFG. I agree with the null pointer check for signalType then. What I don't understand is why a similar code works in 1.5, but it really doesn't matter.

Anyway, the TDB from the original user seems to be quite out of sync. I agree that not crashing is still nicer.